loadtest/loadtest.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package loadtest
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/mattermost/mattermost-load-test-ng/config"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+// LoadTester is a structure holding all the state needed to run a load-test
+type LoadTester struct {
+	controllers   []control.UserController
+	config        *config.LoadTestConfig
+	wg            sync.WaitGroup
+	status        chan control.UserStatus
+	started       bool
+	newController NewController
+}
+
+// NewController is a factory function that returns a new
+// control.UserController given an id and a channel of control.UserStatus
+type NewController func(int, chan<- control.UserStatus) control.UserController
+
+func (lt *LoadTester) handleStatus() {
+	for us := range lt.status {
+		if us.Code == control.USER_STATUS_STOPPED || us.Code == control.USER_STATUS_FAILED {
+			lt.wg.Done()
+		}
+		if us.Code == control.USER_STATUS_ERROR {
+			mlog.Info(us.Err.Error(), mlog.Int("controller_id", us.ControllerId))
+			continue
+		} else if us.Code == control.USER_STATUS_FAILED {
+			mlog.Error(us.Err.Error())
+			continue
+		}
+		mlog.Info(us.Info, mlog.Int("controller_id", us.ControllerId))
+	}
+}
+
+// AddUser increments by one the number of concurrently active users
+func (lt *LoadTester) AddUser() error {
+	if !lt.started {
+		return errors.New("LoadTester is not running")
+	}
+	activeUsers := len(lt.controllers)
+	if activeUsers == lt.config.UsersConfiguration.MaxActiveUsers {
+		return errors.New("Max active users limit reached")
+	}
+	controller := lt.newController(activeUsers+1, lt.status)
+	lt.wg.Add(1)
+	go func() {
+		controller.Run()
+	}()
+	lt.controllers = append(lt.controllers, controller)
+	return nil
+}
+
+// RemoveUser decrements by one the number of concurrently active users
+func (lt *LoadTester) RemoveUser() error {
+	if !lt.started {
+		return errors.New("LoadTester is not running")
+	}
+	activeUsers := len(lt.controllers)
+	if activeUsers == 0 {
+		return errors.New("No active users left")
+	}
+	controller := lt.controllers[activeUsers-1]
```

Comment:
I think to better mimic realistic conditions, it may be better to randomly pop an user from anywhere in the slice. The current method just pops them from the end, which simulates users logging in, and only new users being removed, whereas old users are still there. In a real-world situation, any user might log off.

Reply:
You are right that we want to mimic what happens with real users but at this time my main concern is to not add any extra complexity and especially try to avoid nondeterminism as much as possible. This is the same approach taken with `SimpleController`. It's not a very realistic representation of how a user behaves but it's important to have it nonetheless in order to keep the number of variables down to a minimum and allow us to have more stable results.
I prefer to go back to this kind of details later when we have the whole load-test system up and running and we begin the *tuning* process.
It might be interesting to make the way we remove a user something decidable by the user of this API, for example by passing a typed constant (e.g. random, first, last).

Reply:
Sounds good. Let's document it somewhere (ticket/quick inline comment) so that we don't forget about it later.